In [1]:
#import important lib.
import pandas as pd
import numpy as np

In [2]:
#read excel csv file
data = pd.read_csv('bike_buyers.csv', header = 0)
data.head()

ID Marital Status  Gender   Income  Children        Education  \
0  12496        Married  Female  40000.0       1.0        Bachelors   
1  24107        Married    Male  30000.0       3.0  Partial College   
2  14177        Married    Male  80000.0       5.0  Partial College   
3  24381         Single     NaN  70000.0       0.0        Bachelors   
4  25597         Single    Male  30000.0       0.0        Bachelors   

       Occupation Home Owner  Cars Commute Distance   Region   Age  \
0  Skilled Manual        Yes   0.0        0-1 Miles   Europe  42.0   
1        Clerical        Yes   1.0        0-1 Miles   Europe  43.0   
2    Professional         No   2.0        2-5 Miles   Europe  60.0   
3    Professional        Yes   1.0       5-10 Miles  Pacific  41.0   
4        Clerical         No   0.0        0-1 Miles   Europe  36.0   

  Purchased Bike  
0             No  
1             No  
2             No  
3            Yes  
4            Yes

In [3]:
print(data.dtypes)
print(data.shape)
print(data.describe())

ID                    int64
Marital Status       object
Gender               object
Income              float64
Children            float64
Education            object
Occupation           object
Home Owner           object
Cars                float64
Commute Distance     object
Region               object
Age                 float64
Purchased Bike       object
dtype: object
(1000, 13)
                 ID         Income    Children        Cars         Age
count   1000.000000     994.000000  992.000000  991.000000  992.000000
mean   19965.992000   56267.605634    1.910282    1.455096   44.181452
std     5347.333948   31067.817462    1.626910    1.121755   11.362007
min    11000.000000   10000.000000    0.000000    0.000000   25.000000
25%    15290.750000   30000.000000    0.000000    1.000000   35.000000
50%    19744.000000   60000.000000    2.000000    1.000000   43.000000
75%    24470.750000   70000.000000    3.000000    2.000000   52.000000
max    29447.000000  170000.000000    5.000

In [4]:
data.isnull().sum()

ID                   0
Marital Status       7
Gender              11
Income               6
Children             8
Education            0
Occupation           0
Home Owner           4
Cars                 9
Commute Distance     0
Region               0
Age                  8
Purchased Bike       0
dtype: int64

In [5]:
#copy this file bcz we are drop or clean some data so our real data should not change
df = pd.DataFrame.copy(data)

In [6]:
df

ID Marital Status  Gender    Income  Children        Education  \
0    12496        Married  Female   40000.0       1.0        Bachelors   
1    24107        Married    Male   30000.0       3.0  Partial College   
2    14177        Married    Male   80000.0       5.0  Partial College   
3    24381         Single     NaN   70000.0       0.0        Bachelors   
4    25597         Single    Male   30000.0       0.0        Bachelors   
..     ...            ...     ...       ...       ...              ...   
995  23731        Married    Male   60000.0       2.0      High School   
996  28672         Single    Male   70000.0       4.0  Graduate Degree   
997  11809        Married     NaN   60000.0       2.0        Bachelors   
998  19664         Single    Male  100000.0       3.0        Bachelors   
999  12121         Single    Male   60000.0       3.0      High School   

         Occupation Home Owner  Cars Commute Distance         Region   Age  \
0    Skilled Manual        Yes   0.0        0-1 Miles         Europe  42.0   
1          Clerical        Yes   1.0        0-1 Miles         Europe  43.0   
2      Professional         No   2.0        2-5 Miles         Europe  60.0   
3      Professional        Yes   1.0       5-10 Miles        Pacific  41.0   
4          Clerical         No   0.0        0-1 Miles         Europe  36.0   
..              ...        ...   ...              ...            ...   ...   
995    Professional        Yes   2.0        2-5 Miles  North America  54.0   
996    Professional        Yes   0.0        2-5 Miles  North America  35.0   
997  Skilled Manual        Yes   0.0        0-1 Miles  North America  38.0   
998      Management         No   3.0        1-2 Miles  North America  38.0   
999    Professional        Yes   2.0        10+ Miles  North America  53.0   

    Purchased Bike  
0               No  
1               No  
2               No  
3              Yes  
4              Yes  
..             ...  
995            Yes  
996            Yes  
997            Yes  
998             No  
999            Yes  

[1000 rows x 13 columns]

In [7]:
#column ID in not usefull for further process
df=df.drop(["ID"],axis=1)
df.shape


(1000, 12)

In [8]:
#replace null value
for x in df.columns:
    if df[x].dtype=='object' or df[x].dtype=='bool':
        df[x].fillna(df[x].mode()[0],inplace=True) # some time mode is bimodal there for we want to select 1st mode use mode[0]
    elif df[x].dtype=='int64' or df[x].dtype=='float64':
        df[x].fillna(round(df[x].mean()),inplace=True)

In [9]:
df.shape

(1000, 12)

In [10]:
df.isnull().sum()

Marital Status      0
Gender              0
Income              0
Children            0
Education           0
Occupation          0
Home Owner          0
Cars                0
Commute Distance    0
Region              0
Age                 0
Purchased Bike      0
dtype: int64

In [11]:
colname=[]
for x in df.columns:
    if df[x].dtype=='object':
        colname.append(x)
colname

['Marital Status',
 'Gender',
 'Education',
 'Occupation',
 'Home Owner',
 'Commute Distance',
 'Region',
 'Purchased Bike']

In [19]:
# For preprocessing the data
from sklearn.preprocessing import LabelEncoder

le=LabelEncoder()

for x in colname:
    df[x]=le.fit_transform(df[x])
   


In [20]:
X = df.values[:,0:-1] #select all columns only last column is not consider,
Y = df.values[:,-1] # select only last column,

In [21]:
Y


array([0., 0., 0., 1., 1., 0., 1., 1., 0., 1., 1., 0., 0., 1., 1., 1., 1.,
       0., 1., 1., 1., 0., 1., 0., 0., 0., 1., 0., 0., 1., 0., 1., 0., 1.,
       1., 0., 1., 0., 0., 1., 0., 1., 0., 1., 1., 1., 1., 1., 0., 1., 0.,
       0., 0., 0., 0., 0., 1., 1., 1., 1., 0., 0., 1., 0., 1., 0., 1., 1.,
       1., 0., 1., 0., 0., 1., 0., 0., 0., 1., 1., 1., 1., 0., 1., 0., 1.,
       1., 1., 0., 0., 1., 1., 1., 1., 0., 0., 0., 0., 1., 1., 0., 0., 1.,
       0., 0., 1., 0., 1., 1., 1., 1., 1., 0., 0., 1., 1., 1., 0., 1., 0.,
       0., 1., 0., 0., 0., 1., 0., 0., 0., 1., 1., 0., 1., 1., 1., 0., 0.,
       1., 0., 1., 1., 1., 1., 1., 0., 1., 0., 1., 1., 0., 0., 1., 0., 0.,
       0., 0., 1., 0., 0., 1., 0., 1., 1., 1., 0., 1., 0., 1., 0., 1., 0.,
       1., 0., 0., 0., 1., 1., 1., 0., 1., 1., 0., 1., 0., 1., 0., 1., 1.,
       0., 1., 1., 0., 1., 0., 0., 0., 1., 0., 1., 1., 1., 0., 1., 1., 1.,
       0., 1., 0., 1., 1., 1., 0., 1., 0., 1., 1., 1., 0., 0., 0., 1., 1.,
       0., 0., 0., 0., 0.

In [22]:
X.shape

(1000, 11)

In [23]:
Y.shape

(1000,)

In [24]:
Y=Y.astype(int)

In [25]:
Y

array([0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0,
       1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1,
       1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0,
       1, 1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 0,
       0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 1,
       1, 0, 0, 1, 1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1,
       1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0,
       0, 1, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 1,
       1, 0, 1, 1, 0, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 1, 0, 1,
       1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1,
       1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0,
       1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0,
       1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0,
       0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0,

In [26]:
#scaling the data
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaler.fit(X)
X = scaler.transform(X)
print(X) 

[[-0.91924885 -1.02224741 -0.52545839 ... -1.05375674 -1.28704071
  -0.1927366 ]
 [-0.91924885  0.97823676 -0.84846739 ... -1.05375674 -1.28704071
  -0.10432531]
 [-0.91924885  0.97823676  0.76657763 ...  0.86798807 -1.28704071
   1.39866649]
 ...
 [-0.91924885  0.97823676  0.12055962 ... -1.05375674  0.15583004
  -0.54638172]
 [ 1.08784471  0.97823676  1.41259565 ... -0.41317513  0.15583004
  -0.54638172]
 [ 1.08784471  0.97823676  0.12055962 ...  0.22740647  0.15583004
   0.77978751]]


In [27]:
#split the data in train and test by 80% and 20%

from sklearn.model_selection import train_test_split

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2,
random_state=10)

In [28]:
X_test.shape

(200, 11)

In [29]:
X_train.shape

(800, 11)

# Selecting important features based on Random Forest


In [30]:
from sklearn.ensemble import RandomForestClassifier

model_RandomForest=RandomForestClassifier(n_estimators=50, random_state=10)

#fit the model on the data and predict the values
model_RandomForest.fit(X_train,Y_train)

Y_pred=model_RandomForest.predict(X_test)

In [31]:
from sklearn.metrics import confusion_matrix, accuracy_score,classification_report
#confusion matrix
print(confusion_matrix(Y_test,Y_pred))
print(accuracy_score(Y_test,Y_pred))
print(classification_report(Y_test,Y_pred))

[[86 17]
 [26 71]]
0.785
              precision    recall  f1-score   support

           0       0.77      0.83      0.80       103
           1       0.81      0.73      0.77        97

    accuracy                           0.79       200
   macro avg       0.79      0.78      0.78       200
weighted avg       0.79      0.79      0.78       200



In [32]:
print(list(zip(df.columns[0:-1],model_RandomForest.feature_importances_)))

[('Marital Status', 0.04502702264235507), ('Gender', 0.04777811770216485), ('Income', 0.1386877298990153), ('Children', 0.10465459270365633), ('Education', 0.07044483388380575), ('Occupation', 0.057228576511902315), ('Home Owner', 0.039040484918692775), ('Cars', 0.09128528058884836), ('Commute Distance', 0.09712820733819064), ('Region', 0.054345402936482774), ('Age', 0.2543797508748859)]


# model building for logistic regression.

In [33]:

from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression()
classifier.fit(X_train, Y_train)
Y_pred=classifier.predict(X_test)
print(Y_pred)

[0 1 0 0 1 0 0 0 1 0 0 0 1 1 1 0 0 1 1 0 1 0 0 1 1 0 0 0 0 1 0 0 0 0 0 0 1
 0 0 1 0 0 1 1 0 0 0 1 1 0 1 1 0 1 1 0 0 0 0 1 1 1 0 1 0 0 0 0 0 0 0 0 1 0
 1 0 1 0 0 1 1 0 0 0 0 1 1 0 0 1 0 1 0 0 0 0 0 0 1 0 1 1 1 0 0 0 1 1 1 0 0
 0 0 1 1 1 0 1 1 0 1 0 0 0 1 1 0 1 1 0 1 1 0 0 0 1 0 1 0 1 0 1 1 1 1 1 0 1
 1 0 1 0 0 1 1 0 0 1 0 1 0 0 0 1 0 1 0 0 0 0 1 1 1 0 1 0 0 1 0 0 0 0 0 0 1
 0 0 0 0 0 0 0 0 1 0 1 0 1 1 1]


In [34]:
print(list(zip(Y_test, Y_pred)))

[(0, 0), (1, 1), (0, 0), (0, 0), (1, 1), (0, 0), (0, 0), (0, 0), (0, 1), (1, 0), (0, 0), (1, 0), (1, 1), (1, 1), (0, 1), (0, 0), (0, 0), (1, 1), (1, 1), (0, 0), (1, 1), (1, 0), (1, 0), (0, 1), (0, 1), (1, 0), (0, 0), (1, 0), (0, 0), (1, 1), (0, 0), (0, 0), (0, 0), (0, 0), (1, 0), (0, 0), (0, 1), (0, 0), (0, 0), (0, 1), (1, 0), (0, 0), (0, 1), (1, 1), (1, 0), (0, 0), (0, 0), (1, 1), (1, 1), (1, 0), (0, 1), (1, 1), (0, 0), (1, 1), (1, 1), (1, 0), (0, 0), (1, 0), (0, 0), (1, 1), (1, 1), (1, 1), (0, 0), (1, 1), (0, 0), (0, 0), (1, 0), (1, 0), (0, 0), (0, 0), (1, 0), (0, 0), (1, 1), (0, 0), (1, 1), (0, 0), (1, 1), (1, 0), (0, 0), (1, 1), (1, 1), (0, 0), (0, 0), (1, 0), (1, 0), (0, 1), (1, 1), (0, 0), (0, 0), (1, 1), (1, 0), (1, 1), (1, 0), (1, 0), (0, 0), (1, 0), (0, 0), (1, 0), (1, 1), (0, 0), (1, 1), (0, 1), (0, 1), (1, 0), (0, 0), (0, 0), (0, 1), (1, 1), (1, 1), (0, 0), (0, 0), (0, 0), (0, 0), (0, 1), (1, 1), (1, 1), (0, 0), (0, 1), (1, 1), (1, 0), (1, 1), (0, 0), (1, 0), (0, 0), (0, 1),

In [35]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report

cfm=confusion_matrix(Y_test,Y_pred)
print(cfm)

print("Classification report: ")

print(classification_report(Y_test,Y_pred))

acc=accuracy_score(Y_test, Y_pred)
print("Accuracy of the model: ",acc)

[[79 24]
 [39 58]]
Classification report: 
              precision    recall  f1-score   support

           0       0.67      0.77      0.71       103
           1       0.71      0.60      0.65        97

    accuracy                           0.69       200
   macro avg       0.69      0.68      0.68       200
weighted avg       0.69      0.69      0.68       200

Accuracy of the model:  0.685


In [36]:
Y_pred_prob = classifier.predict_proba(X_test)
print(Y_pred_prob)

[[0.57971795 0.42028205]
 [0.46107965 0.53892035]
 [0.51747945 0.48252055]
 [0.56105861 0.43894139]
 [0.33513442 0.66486558]
 [0.73886098 0.26113902]
 [0.62599841 0.37400159]
 [0.7191676  0.2808324 ]
 [0.47324126 0.52675874]
 [0.602028   0.397972  ]
 [0.53713675 0.46286325]
 [0.58811587 0.41188413]
 [0.3503106  0.6496894 ]
 [0.38375092 0.61624908]
 [0.39389631 0.60610369]
 [0.60303084 0.39696916]
 [0.67969984 0.32030016]
 [0.49650271 0.50349729]
 [0.32559992 0.67440008]
 [0.6286288  0.3713712 ]
 [0.49171381 0.50828619]
 [0.59749818 0.40250182]
 [0.6033681  0.3966319 ]
 [0.38605813 0.61394187]
 [0.43184469 0.56815531]
 [0.58325133 0.41674867]
 [0.66577765 0.33422235]
 [0.72918553 0.27081447]
 [0.71300151 0.28699849]
 [0.26674032 0.73325968]
 [0.69220337 0.30779663]
 [0.6219568  0.3780432 ]
 [0.74254809 0.25745191]
 [0.51531245 0.48468755]
 [0.64593203 0.35406797]
 [0.71105262 0.28894738]
 [0.3577381  0.6422619 ]
 [0.78573541 0.21426459]
 [0.51374127 0.48625873]
 [0.25021357 0.74978643]


In [37]:
Y_pred_class=[]
for value in Y_pred_prob[:,1]:
    if value > 0.49: #after tunning select proper thresold values.
        Y_pred_class.append(1)
    else:
        Y_pred_class.append(0)
print(Y_pred_class)

[0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1]


In [38]:
# GEnerate confusion matrix of logistic regression

from sklearn.metrics import confusion_matrix, accuracy_score, classification_report

cfm=confusion_matrix(Y_test,Y_pred_class)
print(cfm)

print("Classification report: ")

print(classification_report(Y_test,Y_pred_class))

acc=accuracy_score(Y_test, Y_pred_class)
print("Accuracy of the model: ",acc)

[[77 26]
 [36 61]]
Classification report: 
              precision    recall  f1-score   support

           0       0.68      0.75      0.71       103
           1       0.70      0.63      0.66        97

    accuracy                           0.69       200
   macro avg       0.69      0.69      0.69       200
weighted avg       0.69      0.69      0.69       200

Accuracy of the model:  0.69


In [40]:

#find the error for different thresold

for a in np.arange(0.3,0.81,0.01):
    predict_mine = np.where(Y_pred_prob[:,1] > a, 1, 0)
    cfm=confusion_matrix(Y_test, predict_mine)
    total_err=cfm[0,1]+cfm[1,0]
    print("Errors at threshold ", a, ":",total_err, " , type 2 error :",
          cfm[1,0]," , type 1 error:", cfm[0,1])

Errors at threshold  0.3 : 94  , type 2 error : 8  , type 1 error: 86
Errors at threshold  0.31 : 92  , type 2 error : 8  , type 1 error: 84
Errors at threshold  0.32 : 93  , type 2 error : 9  , type 1 error: 84
Errors at threshold  0.33 : 91  , type 2 error : 9  , type 1 error: 82
Errors at threshold  0.34 : 83  , type 2 error : 10  , type 1 error: 73
Errors at threshold  0.35000000000000003 : 83  , type 2 error : 12  , type 1 error: 71
Errors at threshold  0.36000000000000004 : 84  , type 2 error : 15  , type 1 error: 69
Errors at threshold  0.37000000000000005 : 84  , type 2 error : 16  , type 1 error: 68
Errors at threshold  0.38000000000000006 : 74  , type 2 error : 16  , type 1 error: 58
Errors at threshold  0.39000000000000007 : 73  , type 2 error : 17  , type 1 error: 56
Errors at threshold  0.4000000000000001 : 73  , type 2 error : 19  , type 1 error: 54
Errors at threshold  0.4100000000000001 : 73  , type 2 error : 20  , type 1 error: 53
Errors at threshold  0.420000000000000

In [41]:
#predicting using the AdaBoost_Classifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import AdaBoostClassifier

model_AdaBoost=AdaBoostClassifier(base_estimator=DecisionTreeClassifier(random_state=10),
n_estimators=50,
random_state=10)
#fit the model on the data and predict the values
model_AdaBoost.fit(X_train,Y_train)
Y_pred=model_AdaBoost.predict(X_test)

In [42]:
from sklearn.metrics import confusion_matrix, accuracy_score,classification_report
#confusion matrix
print(confusion_matrix(Y_test,Y_pred))
print(accuracy_score(Y_test,Y_pred))
print(classification_report(Y_test,Y_pred))

[[76 27]
 [27 70]]
0.73
              precision    recall  f1-score   support

           0       0.74      0.74      0.74       103
           1       0.72      0.72      0.72        97

    accuracy                           0.73       200
   macro avg       0.73      0.73      0.73       200
weighted avg       0.73      0.73      0.73       200



# predicting using the KNeighbors_Classifier

In [43]:

from sklearn.neighbors import KNeighborsClassifier
model_KNN=KNeighborsClassifier(n_neighbors=int(np.sqrt(len(X_train))), 
metric='minkowski')

model_KNN.fit(X_train,Y_train)

y_pred=model_KNN.predict(X_test)
print(list(zip(Y_test,Y_pred)))

[(0, 0), (1, 1), (0, 0), (0, 1), (1, 1), (0, 1), (0, 1), (0, 0), (0, 0), (1, 1), (0, 0), (1, 1), (1, 1), (1, 1), (0, 1), (0, 0), (0, 0), (1, 0), (1, 0), (0, 0), (1, 1), (1, 0), (1, 1), (0, 0), (0, 1), (1, 1), (0, 0), (1, 0), (0, 0), (1, 1), (0, 0), (0, 0), (0, 0), (0, 1), (1, 0), (0, 1), (0, 1), (0, 0), (0, 0), (0, 0), (1, 1), (0, 1), (0, 0), (1, 1), (1, 1), (0, 0), (0, 0), (1, 1), (1, 1), (1, 1), (0, 0), (1, 0), (0, 0), (1, 0), (1, 1), (1, 1), (0, 0), (1, 0), (0, 0), (1, 1), (1, 0), (1, 1), (0, 0), (1, 1), (0, 1), (0, 0), (1, 1), (1, 1), (0, 0), (0, 0), (1, 1), (0, 1), (1, 1), (0, 0), (1, 1), (0, 0), (1, 0), (1, 1), (0, 1), (1, 1), (1, 0), (0, 0), (0, 0), (1, 0), (1, 0), (0, 1), (1, 1), (0, 0), (0, 0), (1, 0), (1, 0), (1, 1), (1, 1), (1, 0), (0, 0), (1, 1), (0, 0), (1, 0), (1, 0), (0, 0), (1, 1), (0, 0), (0, 0), (1, 1), (0, 0), (0, 0), (0, 1), (1, 1), (1, 1), (0, 0), (0, 1), (0, 0), (0, 0), (0, 1), (1, 1), (1, 1), (0, 0), (0, 1), (1, 1), (1, 1), (1, 1), (0, 1), (1, 0), (0, 1), (0, 1),

In [44]:
#confusion matrix
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report

cfm=confusion_matrix(Y_test,Y_pred)
print(cfm)

print("Classification report: ")

print(classification_report(Y_test,Y_pred))

acc=accuracy_score(Y_test, Y_pred)
print("Accuracy of the model: ",acc)

[[76 27]
 [27 70]]
Classification report: 
              precision    recall  f1-score   support

           0       0.74      0.74      0.74       103
           1       0.72      0.72      0.72        97

    accuracy                           0.73       200
   macro avg       0.73      0.73      0.73       200
weighted avg       0.73      0.73      0.73       200

Accuracy of the model:  0.73


# using SVM

In [45]:
from sklearn.svm import SVC
svc_model=SVC(kernel='rbf',C=1,gamma=0.1)
svc_model.fit(X_train, Y_train)
Y_pred=svc_model.predict(X_test)
print(list(Y_pred))

[0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1]


In [46]:
#generate confusion matrix for svm
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report

cfm=confusion_matrix(Y_test,Y_pred)
print(cfm)

print("Classification report: ")

print(classification_report(Y_test,Y_pred))

acc=accuracy_score(Y_test, Y_pred)
print("Accuracy of the model: ",acc)

[[76 27]
 [34 63]]
Classification report: 
              precision    recall  f1-score   support

           0       0.69      0.74      0.71       103
           1       0.70      0.65      0.67        97

    accuracy                           0.69       200
   macro avg       0.70      0.69      0.69       200
weighted avg       0.70      0.69      0.69       200

Accuracy of the model:  0.695


In [47]:
#in feature selection importance of variable "HOME OWNER" is less compare to other, there we try to drop it and recheck with random forest

In [48]:
df=df.drop(["Home Owner"],axis=1)
df.shape

(1000, 11)

In [49]:
df

Marital Status  Gender    Income  Children  Education  Occupation  Cars  \
0                 0       0   40000.0       1.0          0           4   0.0   
1                 0       1   30000.0       3.0          3           0   1.0   
2                 0       1   80000.0       5.0          3           3   2.0   
3                 1       1   70000.0       0.0          0           3   1.0   
4                 1       1   30000.0       0.0          0           0   0.0   
..              ...     ...       ...       ...        ...         ...   ...   
995               0       1   60000.0       2.0          2           3   2.0   
996               1       1   70000.0       4.0          1           3   0.0   
997               0       1   60000.0       2.0          0           4   0.0   
998               1       1  100000.0       3.0          0           1   3.0   
999               1       1   60000.0       3.0          2           3   2.0   

     Commute Distance  Region   Age  Purchased Bike  
0                   0       0  42.0               0  
1                   0       0  43.0               0  
2                   3       0  60.0               0  
3                   4       2  41.0               1  
4                   0       0  36.0               1  
..                ...     ...   ...             ...  
995                 3       1  54.0               1  
996                 3       1  35.0               1  
997                 0       1  38.0               1  
998                 1       1  38.0               0  
999                 2       1  53.0               1  

[1000 rows x 11 columns]

In [50]:
X = df.values[:,0:-1]
Y = df.values[:,-1]
Y=Y.astype(int)

In [51]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaler.fit(X)
X = scaler.transform(X)
print(X) 

[[-0.91924885 -1.02224741 -0.52545839 ... -1.05375674 -1.28704071
  -0.1927366 ]
 [-0.91924885  0.97823676 -0.84846739 ... -1.05375674 -1.28704071
  -0.10432531]
 [-0.91924885  0.97823676  0.76657763 ...  0.86798807 -1.28704071
   1.39866649]
 ...
 [-0.91924885  0.97823676  0.12055962 ... -1.05375674  0.15583004
  -0.54638172]
 [ 1.08784471  0.97823676  1.41259565 ... -0.41317513  0.15583004
  -0.54638172]
 [ 1.08784471  0.97823676  0.12055962 ...  0.22740647  0.15583004
   0.77978751]]


In [52]:
#split data as train and test
from sklearn.model_selection import train_test_split

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2,
random_state=10)

In [53]:
#after tunning recheck accuracy

model_RandomForest=RandomForestClassifier(n_estimators=50, random_state=10)

#fit the model on the data and predict the values
model_RandomForest.fit(X_train,Y_train)

Y_pred=model_RandomForest.predict(X_test)


from sklearn.metrics import confusion_matrix, accuracy_score,classification_report
#confusion matrix
print(confusion_matrix(Y_test,Y_pred))
print(accuracy_score(Y_test,Y_pred))
print(classification_report(Y_test,Y_pred))

[[78 25]
 [28 69]]
0.735
              precision    recall  f1-score   support

           0       0.74      0.76      0.75       103
           1       0.73      0.71      0.72        97

    accuracy                           0.73       200
   macro avg       0.73      0.73      0.73       200
weighted avg       0.73      0.73      0.73       200



# Result

1. random forest = 78.5%
2. logistic regression = 63%
3. after tunning LR = 69%
4. adaboost regersor = 73%
5. KNN = 69.5%
6. svm = 69.5%
7. After tunning Random forest = 73.5%

so Random forest is best for this data